File: Crypto-Trader-Library/Api-Library/Api-Config/src/main/java/org/cryptotrader/api/library/config/YamlPropertySourceFactory.java
package org.cryptotrader.api.library.config;

import org.jetbrains.annotations.NotNull;
import org.springframework.boot.env.YamlPropertySourceLoader;
import org.springframework.core.env.PropertySource;
import org.springframework.core.env.PropertiesPropertySource;
import org.springframework.core.io.Resource;
import org.springframework.core.io.support.EncodedResource;
import org.springframework.core.io.support.PropertySourceFactory;

import java.io.IOException;
import java.util.List;
import java.util.Properties;

public class YamlPropertySourceFactory implements PropertySourceFactory {
    
    @Override
    public @NotNull PropertySource<?> createPropertySource(String name, EncodedResource encodedResource) throws IOException {
        Resource resource = encodedResource.getResource();
        String sourceName = name != null ? name : resource.getFilename();
        if (sourceName == null) {
            sourceName = "application.yml";
        }
        YamlPropertySourceLoader loader = new YamlPropertySourceLoader();
        List<PropertySource<?>> sources = loader.load(sourceName, resource);
        if (sources.isEmpty()) {
            return new PropertiesPropertySource(sourceName, new Properties());
        }
        return sources.getFirst();
    }
}

File: Crypto-Trader-Library/Security-Library/Security-Config/src/main/kotlin/org/cryptotrader/security/library/config/YamlPropertySourceFactory.kt
package org.cryptotrader.security.library.config

import org.springframework.boot.env.YamlPropertySourceLoader
import org.springframework.core.env.CompositePropertySource
import org.springframework.core.env.PropertySource
import org.springframework.core.io.support.EncodedResource
import org.springframework.core.io.support.PropertySourceFactory

/**
 * Allows using @PropertySource with YAML files.
 */
class YamlPropertySourceFactory : PropertySourceFactory {
    override fun createPropertySource(name: String?, resource: EncodedResource): PropertySource<*> {
        val loader = YamlPropertySourceLoader()
        val filename = resource.resource.filename ?: "application-secure.yml"
        val sources = loader.load(filename, resource.resource)
        if (sources.size == 1) return sources[0]

        val compositeName = name ?: filename
        val composite = CompositePropertySource(compositeName)
        sources.forEach { composite.addPropertySource(it) }
        return composite
    }
}